File: backend/app/routers/journal.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from typing import List
import uuid
import json
from datetime import datetime

from app.models.database import get_session
from app.models.journal import JournalEntry, JournalOutcome
from app.routers.auth import get_default_user
from app.schemas.journal import JournalEntryCreate, JournalEntryRead, JournalEntryUpdate, JournalOutcomeCreate, JournalOutcomeRead

router = APIRouter()


@router.post("", response_model=JournalEntryRead, status_code=201)
async def create_journal_entry(
    entry: JournalEntryCreate,
    session: AsyncSession = Depends(get_session),
):
    user = await get_default_user(session)
    
    db_entry = JournalEntry(
        user_id=user.id,
        scenario_id=entry.scenario_id,
        title=entry.title,
        context=entry.context,
        decision=entry.decision,
        alternatives_considered=json.dumps(entry.alternatives_considered),
        rationale=entry.rationale,
        key_assumptions=json.dumps(entry.key_assumptions),
        success_metrics=json.dumps(entry.success_metrics),
        confidence=entry.confidence,
        review_date=entry.review_date,
    )
    session.add(db_entry)
    await session.flush()
    
    # Re-query with eager-loaded outcomes
    result = await session.execute(
        select(JournalEntry)
        .options(selectinload(JournalEntry.outcomes))
        .where(JournalEntry.id == db_entry.id)
    )
    entry_loaded = result.scalar_one()
    return entry_loaded


@router.get("", response_model=List[JournalEntryRead])
async def list_journal_entries(
    session: AsyncSession = Depends(get_session),
):
    user = await get_default_user(session)
    result = await session.execute(
        select(JournalEntry)
        .options(selectinload(JournalEntry.outcomes))
        .where(JournalEntry.user_id == user.id)
        .order_by(JournalEntry.created_at.desc())
    )
    entries = result.scalars().all()
    return entries


@router.get("/{entry_id}", response_model=JournalEntryRead)
async def get_journal_entry(
    entry_id: uuid.UUID,
    session: AsyncSession = Depends(get_session),
):
    user = await get_default_user(session)
    result = await session.execute(
        select(JournalEntry)
        .options(selectinload(JournalEntry.outcomes))
        .where(JournalEntry.id == entry_id)
        .where(JournalEntry.user_id == user.id)
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    return entry


@router.patch("/{entry_id}", response_model=JournalEntryRead)
async def update_journal_entry(
    entry_id: uuid.UUID,
    entry_update: JournalEntryUpdate,
    session: AsyncSession = Depends(get_session),
):
    user = await get_default_user(session)
    result = await session.execute(
        select(JournalEntry)
        .where(JournalEntry.id == entry_id)
        .where(JournalEntry.user_id == user.id)
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    
    update_data = entry_update.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        if field in ["alternatives_considered", "key_assumptions", "success_metrics"]:
            value = json.dumps(value)
        setattr(entry, field, value)
    
    await session.flush()
    await session.refresh(entry)
    return entry


@router.post("/{entry_id}/outcome", response_model=JournalOutcomeRead, status_code=201)
async def create_journal_outcome(
    entry_id: uuid.UUID,
    outcome: JournalOutcomeCreate,
    session: AsyncSession = Depends(get_session),
):
    user = await get_default_user(session)
    result = await session.execute(
        select(JournalEntry)
        .where(JournalEntry.id == entry_id)
        .where(JournalEntry.user_id == user.id)
    )
    entry = result.scalar_one_or_none()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    
    import json
    db_outcome = JournalOutcome(
        entry_id=entry_id,
        what_happened=outcome.what_happened,
        was_right=outcome.was_right,
        metrics_actual=json.dumps(outcome.metrics_actual),
        what_missed=outcome.what_missed,
        what_got_right=outcome.what_got_right,
        updated_confidence=outcome.updated_confidence,
        lesson=outcome.lesson,
    )
    session.add(db_outcome)
    
    # Update entry
    entry.outcome_captured = True
    entry.outcome_captured_at = datetime.utcnow()
    
    await session.flush()
    await session.refresh(db_outcome)
    return db_outcome

File: frontend/src/lib/user-data/journal.ts
import { ref, set, update, get, remove } from "../firebase"
import type { JournalEntry, JournalOutcome } from "../types"
import { getDb, requireUid, userPath } from "./scope"

function entriesPath(uid: string, entryId?: string): string {
  return entryId
    ? userPath(uid, "journal", "entries", entryId)
    : userPath(uid, "journal", "entries")
}

function outcomePath(uid: string, entryId: string, outcomeId?: string): string {
  return outcomeId
    ? userPath(uid, "journal", "entries", entryId, "outcomes", outcomeId)
    : userPath(uid, "journal", "entries", entryId, "outcomes")
}

export async function loadJournalEntries(): Promise<JournalEntry[]> {
  const database = getDb()
  const uid = requireUid()
  const snap = await get(ref(database, entriesPath(uid)))
  if (!snap.exists()) return []
  const val = snap.val()
  const entries: JournalEntry[] = []
  for (const [id, data] of Object.entries(val)) {
    const d = data as Record<string, unknown>
    if (d.title) {
      const outcomes: JournalOutcome[] = []
      if (d.outcomes && typeof d.outcomes === "object") {
        for (const [, od] of Object.entries(d.outcomes as Record<string, JournalOutcome>)) {
          outcomes.push(od)
        }
      }
      entries.push({
        id,
        user_id: uid,
        title: String(d.title),
        context: String(d.context || ""),
        decision: String(d.decision || ""),
        rationale: String(d.rationale || ""),
        confidence: Number(d.confidence || 5),
        review_date: String(d.review_date || ""),
        outcome_captured: outcomes.length > 0 || Boolean(d.outcome_captured),
        alternatives_considered: (d.alternatives_considered as JournalEntry["alternatives_considered"]) || [],
        key_assumptions: (d.key_assumptions as JournalEntry["key_assumptions"]) || [],
        success_metrics: (d.success_metrics as JournalEntry["success_metrics"]) || [],
        created_at: String(d.created_at || new Date().toISOString()),
        outcomes,
      })
    }
  }
  entries.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())
  return entries
}

export async function createJournalEntry(data: {
  title: string
  context: string
  decision: string
  rationale: string
  confidence: number
  review_date: string
  alternatives_considered: { name: string; description: string }[]
  key_assumptions: { assumption: string; test: string }[]
  success_metrics: { metric: string; target: string }[]
}): Promise<JournalEntry> {
  const database = getDb()
  const uid = requireUid()
  const entryId = crypto.randomUUID()
  const entry: JournalEntry = {
    id: entryId,
    user_id: uid,
    ...data,
    outcome_captured: false,
    created_at: new Date().toISOString(),
  }
  await set(ref(database, entriesPath(uid, entryId)), {
    title: entry.title,
    context: entry.context,
    decision: entry.decision,
    rationale: entry.rationale,
    confidence: entry.confidence,
    review_date: entry.review_date,
    alternatives_considered: entry.alternatives_considered,
    key_assumptions: entry.key_assumptions,
    success_metrics: entry.success_metrics,
    outcome_captured: false,
    created_at: entry.created_at,
  })
  return entry
}

export async function updateJournalEntry(
  entryId: string,
  data: {
    title: string
    context: string
    decision: string
    rationale: string
    confidence: number
    review_date: string
    alternatives_considered: { name: string; description: string }[]
    key_assumptions: { assumption: string; test: string }[]
    success_metrics: { metric: string; target: string }[]
  },
): Promise<void> {
  const database = getDb()
  const uid = requireUid()
  await update(ref(database, entriesPath(uid, entryId)), {
    title: data.title,
    context: data.context,
    decision: data.decision,
    rationale: data.rationale,
    confidence: data.confidence,
    review_date: data.review_date,
    alternatives_considered: data.alternatives_considered,
    key_assumptions: data.key_assumptions,
    success_metrics: data.success_metrics,
  })
}

export async function deleteJournalEntry(entryId: string): Promise<void> {
  const database = getDb()
  const uid = requireUid()
  await remove(ref(database, entriesPath(uid, entryId)))
}

export async function recordOutcome(
  entryId: string,
  data: {
    what_happened: string
    was_right: string
    updated_confidence: number
    lesson: string
  },
): Promise<void> {
  const database = getDb()
  const uid = requireUid()
  const outcomeId = crypto.randomUUID()
  const outcome: JournalOutcome = {
    id: outcomeId,
    ...data,
    what_missed: "",
    what_got_right: "",
    metrics_actual: [],
  }
  await set(ref(database, outcomePath(uid, entryId, outcomeId)), outcome)
  await set(ref(database, `${entriesPath(uid, entryId)}/outcome_captured`), true)
}
